src/compose/service.ts

```diff
@@ -702,24 +725,76 @@ export class Service {
 				log.debug('  Network changes detected');
 			}
 		}
-		return sameNetworks && sameConfig;
+
+		// Check the network mode separetely, as if it is a
+		// service: network mode, the container id needs to be
+		// checked against the running containers
+
+		// When this function is called, it's with the current
+		// state as a parameter and the target as the instance.
+		// We shouldn't rely on that because it's not enforced
+		// anywhere. For that reason we need to consider both
+		// network_modes in the correct way
+		let sameNetworkMode = false;
+		for (const [a, b] of [
+			[this.config.networkMode, service.config.networkMode],
+			[service.config.networkMode, this.config.networkMode],
+		]) {
+			const aMatch = a.match(SERVICE_NETWORK_MODE_REGEX);
+			const bMatch = b.match(SERVICE_NETWORK_MODE_REGEX);
+			if (aMatch != null) {
+				if (bMatch != null) {
+					// They're both service: <name> network modes
+					if (aMatch[1] !== bMatch[1]) {
+						continue;
```

Comment:
With `break`s I suggested, this if construction may be simplified
```typescript
if (aMatch !== null) {
  if (bMatch === null) {
    const containerMatch = b.match(CONTAINER_NETWORK_MODE_REGEX);
    if (containerMatch !== null && currentContainerIds[aMatch[1]] === containerMatch[1]) {
      sameNetworkMode = true;
      break;
    }
  }
} else if (a === b && this.config.hostname === service.config.hostname) {
  sameNetworkMode = true;
  break;
}
```

Reply:
I'm not sure this is any more readable really, I'm not sure what you're trying to optimise for?

Reply:
It's much more readable to me because it contains only branches that we care about to say that configs are equal. There are basically only 2 branches in this code which maps to the actual definition of equality: either (networkMode refers to a service and its container ID matches the config) or (both networkMode and hostname are identical).

Other branches make the code bigger but don't help with deriving this definition from the code.

Reply:
Fair enough, although this is pretty subjective. I'll take a look and see if it can be made clearer.

Reply:
I implemented something similar.